sympy/solvers/solveset.py

```diff
@@ -120,11 +121,63 @@ def _invert_real(f, g_ys, symbol):
             return _invert_real(expo, imageset(Lambda(n, log(n)/log(base)),
                                                g_ys), symbol)
 
+    if isinstance(f, sin) or isinstance(f, csc):
+        n = Dummy('n')
+        if isinstance(g_ys, FiniteSet):
+            fi = sympify('a' + f.func.__name__)
+            # remove all the invalid elements
+            _a = []
+            for i in g_ys:
+                try:
+                    if i.is_real is S.false or \
```

Comment:
I added all these `non-valid elements`(which i called them) because i wanted `solveset_real` to make use of `invert_real` for `trigonometric` instead of `_solve_real_trig`(which is currently used in master) since i thought that would have solve the problem of `n*pi + (-1)**n*asin(y)` thing(one of the test in XFAIL currently)  but after you commented i checked, found my methods would not work.
Thanks for mentioning all important things.